piet-direct2d/src/lib.rs

```diff
@@ -555,20 +559,358 @@ impl<'a> TextLayoutBuilder for D2DTextLayoutBuilder<'a> {
+                    },
+                    metrics: HitTestMetrics {
+                        text_position: http.metrics.text_position() as usize,
+                        is_text: http.metrics.is_text(),
+                    },
+                }
+            })
+    }
+}
+
+/// Counts the number of utf-16 code units in the given string.
+/// from xi-editor
+pub(crate) fn count_utf16(s: &str) -> usize {
+    let mut utf16_count = 0;
+    for &b in s.as_bytes() {
+        if (b as i8) >= -0x40 {
+            utf16_count += 1;
+        }
+        if b >= 0xf0 {
+            utf16_count += 1;
+        }
+    }
+    utf16_count
+}
+
+/// returns utf8 text position (code unit offset)
+/// at the given utf-16 text position
+pub(crate) fn count_until_utf16(s: &str, utf16_text_position: usize) -> Option<usize> {
+    let mut utf8_count = 0;
+    let mut utf16_count = 0;
+    println!("");
+    for &b in s.as_bytes() {
+        if (b as i8) >= -0x40 {
+            utf16_count += 1;
+        }
+        if b >= 0xf0 {
+            utf16_count += 1;
+        }
+
+        utf8_count += 1;
+
+        // When count goes beyond text position, it means the start boundary of the utf16 code unit is passed.
+        // So the utf8 count needs to be backtracked 1.
+        //
+        // char  | utf8 | utf16 | 16_count
+        // é     | 0    | 0     | 0
+        //       | 1    | -     | 1
+        // {0023}| 2    | 1     | 1
+        // {FE0F}| 3    | 2     | 2
+        //       | 4    | -     | 3
+        //       | 5    | -     | 3
+        // {20E3}| 6    | 3     | 3
+        //       | 7    | -     | 4
+        //       | 8    | -     | 4
+        // 1     | 9    | -     | 4
+        if utf16_count == utf16_text_position + 1 {
+            return Some(utf8_count - 1);
+        }
+
+    }
+
+    None
+}
+
+#[cfg(test)]
+mod test {
+    use crate::*;
+    use piet::TextLayout;
+
+    #[test]
+    fn test_hit_test_text_position_basic() {
+        let dwrite = directwrite::factory::Factory::new().unwrap();
+        let mut text_layout = D2DText::new(&dwrite);
+        let font = text_layout.new_font_by_name("Segoe UI", 12.0).build().unwrap();
+
+        let layout = text_layout.new_text_layout(&font, "piet").build().unwrap();
+        let piet_width = layout.width();
+
+        let layout = text_layout.new_text_layout(&font, "pie").build().unwrap();
+        let pie_width = layout.width();
+
+        let layout = text_layout.new_text_layout(&font, "pi").build().unwrap();
+        let pi_width = layout.width();
+
+        let layout = text_layout.new_text_layout(&font, "p").build().unwrap();
+        let p_width = layout.width();
+
+        let layout = text_layout.new_text_layout(&font, "").build().unwrap();
+        let null_width = layout.width();
+
+        let full_layout = text_layout.new_text_layout(&font, "piet text!").build().unwrap();
+
+        assert_eq!(full_layout.hit_test_text_position(3, true).map(|p| p.point.x as f64), Some(piet_width));
```

Comment:
Just to confirm my understanding; because of kerning, substrings may not have the same width when inside a string and when separate from a string.

To implement your suggestion, I'd probably make a macro `assert_kinda_eq` with adjustable tolerance. Let me know if this sounds ok.

Reply:
implemented with an `assert_close_to` test helper fn.